File: src/main/java/com/refinedmods/refinedstorage/container/factory/CrafterManagerContainerProvider.java
package com.refinedmods.refinedstorage.container.factory;

import com.refinedmods.refinedstorage.container.CrafterManagerContainer;
import com.refinedmods.refinedstorage.screen.EmptyScreenInfoProvider;
import com.refinedmods.refinedstorage.tile.CrafterManagerTile;
import net.minecraft.core.BlockPos;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.level.Level;
import net.minecraftforge.items.IItemHandlerModifiable;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Map;

public class CrafterManagerContainerProvider implements MenuProvider {
    private final CrafterManagerTile tile;

    public CrafterManagerContainerProvider(CrafterManagerTile tile) {
        this.tile = tile;
    }

    public static void writeToBuffer(FriendlyByteBuf buf, Level level, BlockPos pos) {
        buf.writeBlockPos(pos);

        Map<Component, List<IItemHandlerModifiable>> containerData = ((CrafterManagerTile) level.getBlockEntity(pos)).getNode().getNetwork().getCraftingManager().getNamedContainers();

        buf.writeInt(containerData.size());

        for (Map.Entry<Component, List<IItemHandlerModifiable>> entry : containerData.entrySet()) {
            buf.writeComponent(entry.getKey());

            int slots = 0;
            for (IItemHandlerModifiable handler : entry.getValue()) {
                slots += handler.getSlots();
            }

            buf.writeInt(slots);
        }
    }

    @Override
    public Component getDisplayName() {
        return new TranslatableComponent("gui.refinedstorage.crafter_manager");
    }

    @Nullable
    @Override
    public AbstractContainerMenu createMenu(int windowId, Inventory playerInventory, Player playerEntity) {
        CrafterManagerContainer container = new CrafterManagerContainer(tile, playerEntity, windowId);

        container.setScreenInfoProvider(new EmptyScreenInfoProvider());
        container.initSlotsServer();

        return container;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/container/factory/CrafterManagerContainerProvider.kt
package com.refinedmods.refinedstorage.container.factory

import com.refinedmods.refinedstorage.api.network.grid.IGrid.slotId
import com.refinedmods.refinedstorage.container.CrafterManagerContainer
import com.refinedmods.refinedstorage.screen.EmptyScreenInfoProvider
import com.refinedmods.refinedstorage.tile.CrafterManagerTile
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.INamedContainerProvider
import net.minecraft.network.PacketByteBuf
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.Text
import net.minecraft.util.text.TranslationTextComponent
import net.minecraft.world.World
import net.minecraftforge.items.IItemHandlerModifiable

class CrafterManagerContainerProvider(private val tile: CrafterManagerTile) : INamedContainerProvider {
    val displayName: Text
        get() = TranslationTextComponent("gui.refinedstorage.crafter_manager")

    @Nullable
    fun createMenu(windowId: Int, playerInventory: PlayerInventory?, playerEntity: PlayerEntity): Container {
        val container = CrafterManagerContainer(tile, playerEntity, windowId)
        container.setScreenInfoProvider(EmptyScreenInfoProvider())
        container.initSlotsServer()
        return container
    }

    companion object {
        fun writeToBuffer(buf: PacketByteBuf, world: World, pos: BlockPos?) {
            buf.writeBlockPos(pos)
            val containerData: Map<Text?, List<IItemHandlerModifiable>> = (world.getBlockEntity(pos) as CrafterManagerTile?).getNode().network.craftingManager.getNamedContainers()
            buf.writeInt(containerData.size)
            for ((key, value) in containerData) {
                buf.writeTextComponent(key)
                var slots = 0
                for (handler in value) {
                    slots += handler.getSlots()
                }
                buf.writeInt(slots)
            }
        }
    }
}